### ISSUE TITLE
How to implement Headless Js config in a expo workflow where there's no index.js file

### ISSUE THREAD

**Issue by LandoSot** [transistorsoft/react-native-background-geolocation#1582]
## Your Environment
* Plugin version: 4.9.3
* Platform: Android
* Device manufacturer / model: Emulator Pixel 4
* React Native version (`react-native -v`): `0.69.6`
* Plugin config

Start / stop functions
```javascript <!-- Syntax-highlighting:  paste your code below -->
function startGeolocation() {
    isBackgroundGeolocationRunning = true
    BackgroundGeolocation.start()
      .then(async () => {
        console.log('- BackgroundGeolocation started')
      })
      .catch((error) => {
        console.log('Error BackgroundGeolocation:', error.message)
      })
  }

  function stopGeolocation() {
    isBackgroundGeolocationRunning = false
    BackgroundGeolocation.stop()
      .then(async () => {
        console.log('- BackgroundGeolocation stoped')
      })
      .catch((error) => {
        console.log('Error BackgroundGeolocation:', error.message)
      })
  }
```

onLocation subscription
```javascript <!-- Syntax-highlighting:  paste your code below -->
const onLocation = BackgroundGeolocation.onLocation(async (location) => {
    // try~catch stuff to my server...
  }, async (error) => {
    console.log('Error in geolocation:', error.message)
  })
```

```javascript <!-- Syntax-highlighting:  paste your code below -->
BackgroundGeolocation.ready({
      activityType: BackgroundGeolocation.ACTIVITY_TYPE_AUTOMOTIVE_NAVIGATION, 
      allowIdenticalLocations: false, 
      backgroundPermissionRationale: { 
        title: 'Permite que Fleter acceda a tu ubicación de fondo',
        message: 'Para poder tener un seguimiento completo del envío, selecciona la opción de "Permitir todo el tiempo"',
        positiveAction: 'Cambiar a "Permitir todo el tiempo"',
        negativeAction: 'Cancelar'
      },
      debug: true,
      desiredAccuracy: BackgroundGeolocation.DESIRED_ACCURACY_HIGH, 
      didDeviceReboot: true, 
      didLaunchInBackground: true, 
      disableElasticity: false, 
      disableMotionActivityUpdates: false, 
      distanceFilter: 50, 
      elasticityMultiplier: 1, 
      enableHeadless: true,
      enableTimestampMeta: false, 
      heartbeatInterval: 60, 
      locationAuthorizationAlert: { 
        titleWhenNotEnabled: 'Permisos de ubicación no consedidos',
        titleWhenOff: 'Es necesario activar el GPS para continuar',
        instructions: 'Selecciona "Todo el tiempo" para el servicio de ubicación',
        cancelButton: 'Cancelar',
        settingsButton: 'Ajustes'
      },
      locationAuthorizationRequest: 'Always', 
      logLevel: BackgroundGeolocation.LOG_LEVEL_INFO,
      notification: { 
        channelName: 'Fleter',
        color: '#228fff',
        priority: BackgroundGeolocation.NOTIFICATION_PRIORITY_MAX,
        sticky: true,
        title: 'Fleter: Rastreo de flete',
        text: 'Rastreo de fondo para los fletes activos de Fleter',
      },
      showsBackgroundLocationIndicator: true,
      startOnBoot: true, 
      stopOnStationary: false,
      stopOnTerminate: false, 
    })
      .then(async (state) => {
        if (!state.enabled)
          startGeolocation()
        console.log("- BackgroundGeolocation is configured and ready: ", state.enabled)
      })
      .catch((error) => {
        console.log('Error al settear BackgroundGeolocation config:', error.message)
      })
```

App.json (main file for an expo workflow project without index.js)
```javascript <!-- Syntax-highlighting:  paste your code below -->
useEffect(() => {
    registerHeadlessTaskAsync()
  }, [])

  async function registerHeadlessTaskAsync() {
    const value = JSON.parse(await AsyncStorage.getItem('isHeadlessTaskRegistered'))
    const isHeadlessTaskRegistered = value || false

    if (!isHeadlessTaskRegistered) {
      console.log('Registering Headless task...')
      BackgroundGeolocation.registerHeadlessTask(taskToHeadless)
      await AsyncStorage.setItem('isHeadlessTaskRegistered', JSON.stringify(true))
    }
  }
```

What to do to keep tracking the onLocation subscription?
```javascript <!-- Syntax-highlighting:  paste your code below -->
export async function taskToHeadless(event) {
  console.log('Event taskToHeadless')
  try {
    let params = event.params;
    console.log('[BackgroundGeolocation HeadlessTask] -', event.name, params)

    switch (event.name) {
    case 'location': {
      console.log('[BackgroundGeolocation HeadlessTask] - onLocation:', location);
      break;
    }
  }
  } catch (error) {
    console.log('taskToHeadless:', error.message)
  }
```

## Expected Behavior

Keep tracking in background even with the screen locked, with the onLocation subscription configuration.

## Actual Behavior

Because the expo workflow project is used, there is no index.js as the main file, making the Headless Js configuration complicated as presented in https://github.com/transistorsoft/react-native-background- geolocation/wiki/Android-Headless-Mode

Said configuration is made in the App.js file, this being the main one, under the use of a useEffect and a function that ensures the registration is done only once.

Due to this the behavior of the tracking system is not as expected, and the tracking stops when the screen is locked

![image](https://user-images.githubusercontent.com/48534682/200375293-8850b22e-1969-4ce5-ad8b-bdc1cc406b13.png)

## Steps to Reproduce

1.
2.
3.
4.

## Context

The application require a tracking system that works all the time until the journey is complete to know the location of the carrier the whole time.

## Debug logs

<details><summary>Logs</summary>

``` <!-- Syntax highlighting:  DO NOT REMOVE -->
PASTE_YOUR_LOGS_HERE
```
![3fab0139-1570-4703-a48b-bb4822d5f5fd](https://user-images.githubusercontent.com/48534682/200374045-2a248236-a580-41b2-9cb9-4de01f0930d1.jpg)

</details>

**Comment by christocracy** [comment]
I wasn’t aware that expo doesn’t have an `index.js`.  I will investigate.

**Comment by LandoSot** [comment]
Extra question:
With the headless js task, is it possible to keep tracking the device with the onLocation subscription?
Or only with the heartbeat event getting the current position?

This is because we prefer to track the device by distance with "distanceFilter" like the ready configuration in onLocation subscription, and as I understand, heartbeat pulses are by time.

**Comment by christocracy** [comment]
>is it possible to keep tracking the device with the onLocation subscription

No.  There is no `onLocation` (or any other `onXXX` event-listener when your app is terminated.

This is why Headless mode exists.  All events are directed to your registered headless-task.

Providing a headless-task or not does **not** affect the plugin's operation.  As long as `stopOnTerminate: false`, the plugin *will* continue tracking.

Providing a headless-task is only required for if you have business-logic to perform in the terminated state.

The plugin is perfectly capable of continuing to upload locations to your server with its built-in HTTP service if you provide a `Config.url`.

**Comment by christocracy** [comment]
regarding your original issue, looking at `package.json` of a freshly generated app:

```json
{
  "name": "my-app",
  "version": "1.0.0",
  "main": "node_modules/expo/AppEntry.js"
}
```

You might be able to simply create your own `index.js`, then copy the original contents of `node_modules/expo/AppEntry.js` into it, then add the prescribed headless-task according to the docs there.

```diff
{
  "name": "my-app",
  "version": "1.0.0",
+  "main": "index.js"
}
```

*index.js*
```javascript
import registerRootComponent from 'expo/build/launch/registerRootComponent';

// Import BackgroundGeolocation
import BackgroundGeolocation from "react-native-background-geolocation";

import App from '../../App';

registerRootComponent(App);

const HeadlessTask = async (event) => {
  let params = event.params;
  console.log('[BackgroundGeolocation HeadlessTask] -', event.name, params);  
}

////
// Register your HeadlessTask with BackgroundGeolocation plugin.
//
BackgroundGeolocation.registerHeadlessTask(HeadlessTask);
```

**Comment by christocracy** [comment]
And `index.js` could *possibly* be simplified like so:

```javascript
import './node_modules/expo/AppEntry'

// Import BackgroundGeolocation
import BackgroundGeolocation from "react-native-background-geolocation";

const HeadlessTask = async (event) => {
  let params = event.params;
  console.log('[BackgroundGeolocation HeadlessTask] -', event.name, params);  
}

////
// Register your HeadlessTask with BackgroundGeolocation plugin.
//
BackgroundGeolocation.registerHeadlessTask(HeadlessTask);
```
```

**Comment by LandoSot** [comment]
Then my issue is that when I lock the screen, the onLocation subscription stops executing its process.

In essence, this is what my onLocation do
```javascript
const onLocation = BackgroundGeolocation.onLocation(async (location) => {
  //  Get required object params from AsyncStorage
  const dataArray = JSON.parse(await AsyncStorage.getItem('dataArray'))
  let request = []

  // map the array and generate custom objects including latitude and longitude
  dataArray.map(item => {
    newCoords = {
      dateTime: `${moment(new Date()).format('YYYY-MM-DD HH:mm')}`,
      idShipment: item.idShipment,
      latitude: location.coords.latitude,
      longitude: location.coords.longitude
    }

    // Add the object to the array for request
    request.push(newCoords)
  )}

  // Custom POST of request array to server 
  // ...
}, async (error) => {
    console.log('Error in geolocation:', error.message)
})
```

This process works fine all the time while the screen is unlocked (on), but if the screen is locked by de user or timeout, the onLocation process stops and it didn't send any other data to our server.

Once the screen is turned on again (unlocked) the process still works normally.

**Comment by christocracy** [comment]
Why don’t you let the plugin take care of your http?

**Comment by LandoSot** [comment]
The app should work offline, so when any query (post, put) fails, the whole flow of the app needs to still work.

In each HTTP query, I have processes of what to do when the user hasn't have internet or the query fails.
Also, I have an extra step in each query to check for pending information to sync, before executing the post/put.

**Comment by christocracy** [comment]
In the demo app, I add a location marker to the map using onLocation listener.

My devices are always running in the background and location markers are always drawn on the map.

**Comment by LandoSot** [comment]
Yes, we test your demo before buying and we notice that everything works fine, and that's why we can't understand why it doesn't work with the screen locked.

Even I include a function to save in AsyncStorage a kind of log for each time that onLocation executes, and as I said, everything works with the screen unlocked, but with the screen locked I don't get records in my AsyncStorage log, evidence that the subscription is not executing.

In the beginning, I thought that was because I wasn't registering a task in Headless JS, but if it isn't required for subscriptions, I don't know what is missing.

Is it because I'm using expo?
Maybe the background subscriptions only work with native projects?

**Comment by christocracy** [comment]
> Is it because I'm using expo?

no

>Maybe the background subscriptions only work with native projects?

React Native is not a “native project”.

When all is said and done, Expo is nothing more than the same old RN app that happens to get built on a remote server instead of your localhost.

see api docs *Logger*.  Push some log entries into the plugin log database.

You can fetch your logs via .emailLog method.

**Comment by LandoSot** [comment]
Hope this can help
The only active subscription is onLocation

![Log 01](https://user-images.githubusercontent.com/48534682/200930894-dfb83aeb-2e3a-48c1-92bc-c27ae4a44e02.jpeg)
![Log 02](https://user-images.githubusercontent.com/48534682/200930680-cb6f2a0a-7748-4409-8d5c-a00e97613b31.jpeg)
![Log 03](https://user-images.githubusercontent.com/48534682/200930744-3106794b-4e79-42f0-bb8b-f3ce21c1d377.jpeg)
![Log 04](https://user-images.githubusercontent.com/48534682/200930775-02163ccd-c308-464f-a836-32574b1aee62.jpeg)
![Log 05](https://user-images.githubusercontent.com/48534682/200930784-13125828-fa46-454b-927c-b17d85abec7d.jpeg)
![Log 06](https://user-images.githubusercontent.com/48534682/200930816-5256a94a-da86-4772-b605-d7717a26c36c.jpeg)
![Log 07](https://user-images.githubusercontent.com/48534682/200930825-54115fde-0359-4afc-b8ec-5d740279bbf7.jpeg)
![Log 08](https://user-images.githubusercontent.com/48534682/200930836-be8d3d42-ac87-4e7a-b62d-f6d84866e386.jpeg)
![Log 09](https://user-images.githubusercontent.com/48534682/200930856-c10ab2f9-eee2-4c54-a181-53630289fd83.jpeg)
![Log 10](https://user-images.githubusercontent.com/48534682/200930867-ebdc037d-fbed-4d2c-9307-ab410ed072da.jpeg)


We still generating logs, in a few minutes I'll share more.

**Comment by christocracy** [comment]
I don’t want to see a stream of screenshots.

I want the log file

**Comment by LandoSot** [comment]
### Here's the log

[background-geolocation.log.gz](https://github.com/transistorsoft/react-native-background-geolocation/files/9982669/background-geolocation.log.gz)

- Between `11-09 18:35:49` and `11-09 19:25:07` the app was used with the screen looked, but the onLocation process stopped and we don't get any sync of coordinates.
- We notice more skipping time events at `11-09 19:46:14`, and another one at `11-09 23:55:44`.
- The plugin looks like it stops itself at `11-10 03:19:27` (in the night) and continues at `11-10 07:16:53`, but the onLocation process didn't share any other location.

**Comment by LandoSot** [comment]
### Hi, we still waiting for an answer.
Feel free to ask for information or configuration extra if required.

**Comment by christocracy** [comment]
I see no explanation for no locations being recorded between `18:35` and `19:25`.

I suggest you see https://dontkillmyapp.com and test on a variety of devices.

**Comment by rodrigodiasf1984** [comment]
> I wasn’t aware that expo doesn’t have an `index.js`. I will investigate.

I'm using expo bare workflow and I´ve got an index.js  however in managed work there is no index.js.

bare workflow
![image](https://user-images.githubusercontent.com/30758309/202511790-d6f89487-cb20-44be-9c01-ba64b30b8a50.png)

managed workflow
![image](https://user-images.githubusercontent.com/30758309/202512018-f528875a-6fd7-4dd8-9fc0-a5dabcce095a.png)

**Comment by christocracy** [comment]
You can create your own `index.js`.

```javascript
import { registerRootComponent } from 'expo';
import App from './App';

registerRootComponent(App);
```

Then delete this line from your `package.json`:

```diff
{
-  "main": "node_modules/expo/AppEntry.js",
}
```

**Comment by christocracy** [comment]
Here is a reference to this in the [Expo Docs](https://docs.expo.dev/guides/monorepos/#change-default-entrypoint)

<img width="1222" alt="Screen Shot 2022-11-17 at 12 29 33 PM" src="https://user-images.githubusercontent.com/14999/202516231-eda243c7-2a4f-4f3c-82bc-8f76ab1632d9.png">
